### ISSUE TITLE
How to use this example using Canvas as a platform?

### ISSUE THREAD

**Issue by mwleinad** [IMSGlobal/lti-1-3-php-library#23]
I've been reading about the specification and still trying to figure things out.
But for now, at least, the tool is working using my own imsglobal platform
https://lti-ri.imsglobal.org/platforms/395/

But now, do you know how to setup this in a different platform? I'm using Canvas. 
When I launch the tool, it goes to game_example/login.php using this data

https://lti-ri.imsglobal.org/platforms/395/authorizations/new?
scope=openid
response_type=id_token
response_mode=form_post
prompt=none
client_id=Cloud
redirect_uri=http%3A%2F%2Flocalhost%3A9001%2Fgame_example%2Fgame.php
state=state-5d541bcf323c28_11275570
nonce=nonce-d541bcf323ee3.44777709
login_hint=374a6c8841a87aa700dcf83b144106fc4e2c86aa
lti_message_hint=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ2ZXJpZmllciI6IjVkMGY1NWQwZDZhODQyY2E5NDIwMmIyZjRjZTViZjczODI2MTBhYmUyYWQ0ZjQ0NDk2OTZmYmI1YjEyYmI5YTRjNjQ0ZWQ4ZGU1YzZmMzQ0MTM3ODZkZjU0ZDJjYzNlMjE5ZDAzMzFmNzdiM2Y5N2E5ZDJiYzczOTM3MWQ3Y2Q3IiwiY2FudmFzX2RvbWFpbiI6InZpZGdyaWQuaW5zdHJ1Y3R1cmUuY29tIiwiY29udGV4dF90eXBlIjoiQ291cnNlIiwiY29udGV4dF9pZCI6MTQ5NzEwMDAwMDAwMDAwMDM5LCJleHAiOjE1NjU3OTM0MTh9.su5kb18R_qv74nkatejsbRVlF1hkE-KH4E5U5Ewdbek

But 
https://lti-ri.imsglobal.org/platforms/395/authorizations/new
Fails with a 500 error.

This is what I have in my local.json

    "https://canvas.instructure.com": {
        "client_id": "Cloud",
        "auth_login_url": "https://lti-ri.imsglobal.org/platforms/395/authorizations/new",
        "auth_token_url": "https://lti-ri.imsglobal.org/platforms/395/access_tokens",
        "key_set_url": "https://lti-ri.imsglobal.org/platforms/395/platform_keys/389.json",
        "private_key_file": "/private.key",
        "deployment": {
            "1": "1"
        }
    },

I'm PRETTY sure I'm missing something, but I'm at my wits end by now.

Thanks.

**Comment by mwleinad** [comment]
Found the solution yesterday
in my local.json
    "https://canvas.instructure.com": {
        "client_id": "//Api key id key from canvas
        "auth_login_url": //Canvas auth login url,
        "auth_token_url": //Still trying to figure out this one
        "key_set_url": //Canvas url to public keys
        "private_key_file": //Tool private key
        "deployment": {
            "deplymentIdFromCanvas": "deplymentIdFromCanvas" //This comes from canvas too.
        }
    },

Questions:

- Everything except the private_key_file comes from the Platform (In this case Canvas)??
- How do I get the deploymentId? I was able to get it by reading this variable
`$this->jwt['body']['https://purl.imsglobal.org/spec/lti/claim/deployment_id']`
and applying the value to local.json.
But do I need to do that everytime?

Now for the fun part, deep linking :/

**Comment by MartinLenord** [comment]
Hey @mwleinad,
Canvas have some unofficial documentation on 1.3 here https://docs.google.com/document/d/1g2ODyuXvYGYfpZHwxBOcjqZOJwjme_85t72F3jbWb4E/edit i don't think there are any official docs for it yet

**Comment by mwleinad** [comment]
Yeah, it was very hard to figure out, but I finally was able to 😄 
I still don't know how I'm supposed to get the deployment id though.

Maybe you can point me in the right direction with something else?
How can I get the email for the person that launched the tool?

Thanks!

**Comment by MartinLenord** [comment]
You can have a look in the id_token JWT that is passed in the launch, the deployment_id will be in there under the `https://purl.imsglobal.org/spec/lti/claim/deployment_id` claim

You can get the email address `$launch->get_launch_data()['email']` which gets the `email` claim in the `id_token`

**Comment by mwleinad** [comment]
What if the launch data doesn't not contain the email?

```
Array
(
    [https://purl.imsglobal.org/spec/lti/claim/message_type] => LtiDeepLinkingRequest
    [https://purl.imsglobal.org/spec/lti/claim/version] => 1.3.0
    [https://purl.imsglobal.org/spec/lti-dl/claim/deep_linking_settings] => Array
        (
            [deep_link_return_url] => https://vidgrid.instructure.com/courses/39/deep_linking_response?modal=true
            [accept_types] => Array
                (
                    [0] => link
                    [1] => file
                    [2] => html
                    [3] => ltiResourceLink
                    [4] => image
                )

            [accept_presentation_document_targets] => Array
                (
                    [0] => embed
                    [1] => iframe
                    [2] => window
                )

            [accept_media_types] => image/*,text/html,application/vnd.ims.lti.v1.ltilink,*/*
            [accept_multiple] => 
            [auto_create] => 
            [validation_context] => 
            [errors] => Array
                (
                    [errors] => Array
                        (
                        )

                )

        )

    [aud] => 149710000000000004
    [azp] => 149710000000000004
    [https://purl.imsglobal.org/spec/lti/claim/deployment_id] => 77:ef42577560dbb9f4932b829b1ebcaab31428040b
    [exp] => 1566228464
    [iat] => 1566224864
    [iss] => https://canvas.instructure.com
    [nonce] => nonce-5d5ab1e1340402.37906781
    [sub] => bf1e68b3-5b2c-4f7b-9be6-5489bf890aaf
    [https://purl.imsglobal.org/spec/lti/claim/target_link_uri] => http://localhost:9001/game_example/game.php
    [https://purl.imsglobal.org/spec/lti/claim/context] => Array
        (
            [id] => ef42577560dbb9f4932b829b1ebcaab31428040b
            [label] => CloudVidgrid
            [title] => CloudVidgrid
            [type] => Array
                (
                    [0] => http://purl.imsglobal.org/vocab/lis/v2/course#CourseOffering
                )

            [validation_context] => 
            [errors] => Array
                (
                    [errors] => Array
                        (
                        )

                )

        )

    [https://purl.imsglobal.org/spec/lti/claim/tool_platform] => Array
        (
            [guid] => NT9pgxz1jHdyoPOyqagEKwoPZkQAl87ZbN5nzCqU:canvas-lms
            [name] => VidGrid
            [version] => cloud
            [product_family_code] => canvas
            [validation_context] => 
            [errors] => Array
                (
                    [errors] => Array
                        (
                        )

                )

        )

    [https://purl.imsglobal.org/spec/lti/claim/launch_presentation] => Array
        (
            [document_target] => iframe
            [height] => 400
            [width] => 800
            [return_url] => https://vidgrid.instructure.com/courses/39/external_content/success/external_tool_dialog
            [locale] => en
            [validation_context] => 
            [errors] => Array
                (
                    [errors] => Array
                        (
                        )

                )

        )

    [locale] => en
    [https://purl.imsglobal.org/spec/lti/claim/roles] => Array
        (
            [0] => http://purl.imsglobal.org/vocab/lis/v2/institution/person#Administrator
            [1] => http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor
            [2] => http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor
            [3] => http://purl.imsglobal.org/vocab/lis/v2/system/person#User
        )

    [https://purl.imsglobal.org/spec/lti/claim/custom] => Array
        (
        )

    [errors] => Array
        (
            [errors] => Array
                (
                )

        )

)

```

I'm taking a look at Canvas settings but I can't find one to "send the email" like in lti 1.1. No luck so far

**Comment by MartinLenord** [comment]
Email is not a required field, so the platform does not have to send it. I believe in Canvas when you create your tool credentials there is an option to set the privacy level that determines if emails are sent

**Comment by viveklm** [comment]
> * Everything except the private_key_file comes from the Platform (In this case Canvas)??

Getting following error not sure what value and where to set private_key_file??

{"errors":{"jwt":[{"attribute":"jwt","type":"JWT verification failure","message":"JWT verification failure"}]}}

**Comment by viveklm** [comment]
> I still don't know how I'm supposed to get the deployment id though.

Deployment Id will get from External Apps. for ease of reference check screenshot below.

![image](https://user-images.githubusercontent.com/31856032/64917050-9b5aef80-d759-11e9-8ef3-ba80b21af5e5.png)

**Comment by jd555** [comment]
Did you determine what the auth_token_url should be? I found some Canvas oAuth 2.0 docs (https://canvas.instructure.com/doc/api/file.oauth_endpoints.html) that said it would be /login/oauth2/token though I get a page not found with my own open source Canvas set up. Plus it also wants a "client_secret", but does this library support that?

**Comment by vivekaswal** [comment]
> Yeah, it was very hard to figure out, but I finally was able to 
> I still don't know how I'm supposed to get the deployment id though.
> 
> Maybe you can point me in the right direction with something else?
> How can I get the email for the person that launched the tool?
> 
> Thanks!

Deeplinking service https://xxxxx.instructure.com/courses/175/deep_linking_response?modal=true giving 500.  Is there anything which I'm missing?

![image](https://user-images.githubusercontent.com/10030481/75820016-9884a300-5d69-11ea-8570-ba0904638e6d.png)
